packages/react-admin/src/AdminRoutes.js

```diff
@@ -1,71 +1,103 @@
-import React, { Children, Component, cloneElement } from 'react';
+import React, { Children, Component, cloneElement, createElement } from 'react';
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
 import { Redirect, Route, Switch } from 'react-router-dom';
 import compose from 'recompose/compose';
 import getContext from 'recompose/getContext';
 
+import { declareResources as declareResourcesAction } from './actions';
 import Loading from './mui/layout/Loading';
 import NotFound from './mui/layout/NotFound';
 import WithPermissions from './auth/WithPermissions';
 import { AUTH_GET_PERMISSIONS } from './auth/types';
-import { isLoggedIn } from './reducer';
+import { isLoggedIn, getResources } from './reducer';
+import DefaultLayout from './mui/layout/Layout';
+import Logout from './mui/auth/Logout';
+import Menu from './mui/layout/Menu';
 
 export class AdminRoutes extends Component {
-    state = { childrenToRender: [] };
+    state = { children: [] };
 
     componentWillMount() {
         this.initializeResources(this.props);
     }
 
-    initializeResources = nextProps => {
-        if (typeof nextProps.children === 'function') {
-            this.initializeResourcesAsync(nextProps);
+    initializeResources = props => {
+        if (typeof props.children === 'function') {
+            props.authClient(AUTH_GET_PERMISSIONS).then(permissions => {
```

Comment:
There is no risk that resources won't be declared when custom routes will be mounted?

Reply:
The e2e tests confirm that yes :)